### PR TITLE
Feature: Add ArgoCD widget

### PR DESCRIPTION
**Pull request by fcornelius**
## Proposed change

This adds a widget for [ArgoCD](https://argo-cd.readthedocs.io/en/latest/). It can list the count of apps present in an ArgoCD instance by app status, e.g. `synced` or `healthy`.

image

Closes #2644

## Type of change

- [x] New service widget
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation only
- [ ] Other (please explain)

## Checklist:

- [x] If applicable, I have added corresponding documentation changes.
- [x] If applicable, I have reviewed the [feature](https://gethomepage.dev/more/development/#new-feature-guidelines) and / or [service widget guidelines](https://gethomepage.dev/more/development/#service-widget-guidelines).
- [x] I have checked that all code style checks pass using [pre-commit hooks](https://gethomepage.dev/more/development/#code-formatting-with-pre-commit-hooks) and [linting checks](https://gethomepage.dev/more/development/#code-linting).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.

### CODE REVIEW

**Review by shamoon — CHANGES_REQUESTED** [review]
The length of the docs here backs up my theory that k8s is too complicated. If you’d think there’s a way to make it more concise (mostly linking to other places ) that’d be great, I didn’t see anything obvious.

As for the widget, if you wouldn’t mind copy pasting some sample api output for me to test.

Finally, one thing is all widgets should display no more than 4 blocks. I’ll have a real look once you have a chance to supply the above.

thanks

**Comment by fcornelius** [comment]
I feel you, the doc however is not related to kubernetes itself. It's due to the way roles are configured in ArgoCD, which is a little unintuitive and I wanted to include a security best practice rather than just the quickest solution to get it running.

Limited the maximum fields shown to 4 now.

Sample API response to test with: [argocd-api-response.json](https://github.com/user-attachments/files/17820774/argocd-api-response.json)
This one should yield the fields:
`apps`: 4, `synced`: 2, `outOfSync`: 2, `healthy`: 1, `missing`: 1, `degraded`: 1, `progressing`: 1

File: src/widgets/prometheusmetric/component.jsx [hunk 1]
```diff
@@ -5,6 +5,7 @@ import Block from "components/services/widget/block";
 import useWidgetAPI from "utils/proxy/use-widget-api";
 
 function formatValue(t, metric, rawValue) {
+  if (!metric?.format) return rawValue;
```
Comment (shamoon) [inline src/widgets/prometheusmetric/component.jsx:8]: Thanks, why is this here btw?

File: src/widgets/argocd/component.jsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import useWidgetAPI from "utils/proxy/use-widget-api";
+
+export default function Component({ service }) {
+  const { widget } = service;
+
+  // Limits fields to available statuses
+  const validFields = ["apps", "synced", "outOfSync", "healthy", "progressing", "degraded", "suspended", "missing"];
+  widget.fields = widget.fields?.filter((field) => validFields.includes(field));
+
+  // Limits max number of displayed fields
```
Comment (shamoon) [inline src/widgets/argocd/component.jsx:12]:
```suggestion
  if (!widget.fields) {
    widget.fields = ["apps", "synced", "outOfSync", "healthy"];
  }

```

1. we dont do this kind of field validation here
2. At the moment the widget doesn't work if the user doesn't supply any fields, which is pretty common

File: src/widgets/argocd/component.jsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+  const { data: appsData, error: appsError } = useWidgetAPI(widget, "applications");
+
+  const appCounts = widget.fields?.map((status) => {
+    if (status === "apps") {
+      return { status, count: appsData?.items?.length };
+    }
+    const apiStatus = status.charAt(0).toUpperCase() + status.slice(1);
+    const count = appsData?.items?.filter(
+      (item) => item.status?.sync?.status === apiStatus || item.status?.health?.status === apiStatus,
```
Comment (shamoon) [inline src/widgets/argocd/component.jsx:26]:
```suggestion
    const count = appsData?.items?.filter(
      (item) => item.status?.sync?.status.toLowerCase() === status.toLowerCase() || item.status?.health?.status.toLowerCase() === status.toLowerCase(),
```

File: public/locales/en/common.json [hunk 4]
```diff
@@ -988,5 +988,15 @@
       "memory": "MEM",
       "disk": "Disk",
       "network": "NET"
+    },
+    "argocd": {
+        "apps": "Apps",
+        "synced": "Synced",
+        "outOfSync": "OutOfSync",
```
Comment (shamoon) [inline public/locales/en/common.json:995]:
```suggestion
        "outOfSync": "Out Of Sync",
```

**Review by shamoon — CHANGES_REQUESTED** [review]
Thanks, seems to work well just some small things.

Screenshot 2024-11-19 at 12 50 03 PM

File: src/widgets/argocd/component.jsx (hunk 2 above)
Comment (fcornelius) [reply src/widgets/argocd/component.jsx:12]: Cool, good to know

File: src/widgets/prometheusmetric/component.jsx (hunk 1 above)
Comment (shamoon) [reply src/widgets/prometheusmetric/component.jsx:8]: Oh, I didn't realize you're the same person, cool that's fine

Comment (fcornelius) [reply src/widgets/prometheusmetric/component.jsx:8]: No problem 😄

**Review by shamoon — APPROVED** [review]
Awesome thanks

**Comment by fcornelius** [comment]
> Awesome thanks

Thank you for this cool project!

**Pull request merged**